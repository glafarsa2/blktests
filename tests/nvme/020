#!/bin/bash
# Test NVMe DSM Discard command on NVMeOF with a file-backed ns.
# Copyright (c) 2017-2018 Western Digital Corporation or its affiliates.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA  02110-1301, USA.
#
#   Author: Chaitanya Kulkarni <chaitanya.kulkarni@wdc.com>
#

. tests/nvme/rc

DESCRIPTION="test NVMe DSM Discard command on NVMeOF file-backed ns"
QUICK=1

requires() {
	_have_program nvme && _have_module nvme-loop && _have_configfs \
		_have_module nvmet
}

test() {
	echo "Running ${TEST_NAME}"

	modprobe nvmet
	modprobe nvme-loop

	local port
	local nvmedev
	local file_path="$TMPDIR/img"
	local subsys_name="blktests-subsystem-1"
	local nblk_range="10,10,10,10,10,10,10,10,10,10"
	local sblk_range="100,200,300,400,500,600,700,800,900,1000"

	truncate -s 1G "${file_path}"

	_create_nvmet_subsystem "${subsys_name}" "${file_path}" \
		"91fdba0d-f87b-4c25-b80f-db7be1418b9e"
	port="$(_create_nvmet_port "loop")"
	_add_nvmet_subsys_to_port "${port}" "${subsys_name}"

	nvme connect -t loop -n "${subsys_name}"

	nvmedev="$(_find_nvme_loop_dev)"
	cat "/sys/block/${nvmedev}n1/uuid"
	cat "/sys/block/${nvmedev}n1/wwid"

	nvme dsm "/dev/${nvmedev}" -n 1 -d -s "${sblk_range}" -b "${nblk_range}"

	nvme disconnect -n "${subsys_name}"

	_remove_nvmet_subsystem_from_port "${port}" "${subsys_name}"
	_remove_nvmet_subsystem "${subsys_name}"
	_remove_nvmet_port "${port}"

	rm -f "${file_path}"

	modprobe -r nvme-loop
	modprobe -r nvmet
	echo "Test complete"
}
