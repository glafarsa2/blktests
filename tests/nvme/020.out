Running nvme/020
91fdba0d-f87b-4c25-b80f-db7be1418b9e
uuid.91fdba0d-f87b-4c25-b80f-db7be1418b9e
NVMe DSM: success
NQN:blktests-subsystem-1 disconnected 1 controller(s)
Test complete
